### PR TITLE
Add PATCH /ingest/:id to ingest tracker

### PR DESCRIPTION
**Pull request by kenoir**
Accepts `IngestUpdate` as request.

Part of https://github.com/wellcomecollection/platform/issues/4419

### CODE REVIEW

File: ingests_tracker/src/main/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApi.scala [hunk 1]
```diff
@@ -44,13 +44,21 @@ trait IngestsTrackerApi extends Runnable with Logging {
           }
         }
       },
+      patch {
+        pathPrefix("ingest" / JavaUUID) { id =>
+          entity(as[IngestUpdate]) { ingestUpdate =>
+            info(s"${id}: ${ingestUpdate}")
```
Comment (alexwlchan) [inline ingests_tracker/src/main/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApi.scala:50]: nit:

```suggestion
            info(s"$id: $ingestUpdate")
```

File: ingests_tracker/src/main/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApi.scala [hunk 2]
```diff
@@ -44,13 +44,21 @@ trait IngestsTrackerApi extends Runnable with Logging {
...
+            complete(StatusCodes.InternalServerError)
```
Comment (alexwlchan) [inline ingests_tracker/src/main/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApi.scala:51]: 🤔

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala [hunk 3]
```diff
@@ -91,6 +94,43 @@ class IngestsTrackerApiFeatureTest
...
+      createIngestStatusUpdateWith(
+        id = ingest.id,
+        status = Succeeded
+      )
+
+    // Cases:
+    // - Updating existing ingest 200
+    // - Updating existing ingest with bad data 500
+    // - Updating non-existent ingest 500
+    // - Broken app 500
```
Comment (alexwlchan) [inline ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:116]: If you look at the IngestUpdater state machine logic, it's possible for it to reject an ingest update (e.g. if a previously Failed ingest gets a status update "Succeeded"). Maybe that's a 400, not a 500?

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala [hunk 4]
```diff
@@ -91,6 +94,43 @@ class IngestsTrackerApiFeatureTest
...
+    val ingestStatusUpdate =
+      createIngestStatusUpdateWith(
+        id = ingest.id,
+        status = Succeeded
+      )
+
+    // Cases:
+    // - Updating existing ingest 200
+    // - Updating existing ingest with bad data 500
+    // - Updating non-existent ingest 500
```
Comment (alexwlchan) [inline ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:115]: 404 surely?

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala (hunk 3 above)
Comment (kenoir) [reply ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:126]: I think probably 409. I'm going to do:

```
    // - Updating existing ingest with bad json 400
    // - Updating existing ingest to incorrect state 409
```

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala (hunk 4 above)
Comment (kenoir) [reply ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:125]: 409 again I think.

> "The HTTP 409 Conflict response status code indicates a request conflict with current state of the server."

File: ingests_tracker/src/main/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApi.scala (hunk 2 above)
Comment (kenoir) [reply ingests_tracker/src/main/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApi.scala]: PR is incomplete.

**Comment by kenoir** [comment]
The next PR will add the `sendOngoingMessages` method from the current ingests service.

File: common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/models/IngestUpdate.scala [hunk 5]
```diff
@@ -5,21 +5,27 @@ import uk.ac.wellcome.platform.archive.common.bagit.models.BagVersion
...
   val events: Seq[IngestEvent]
+  val `type`: String
 }
 
-case class IngestEventUpdate(id: IngestID, events: Seq[IngestEvent])
-    extends IngestUpdate
+case class IngestEventUpdate(
+  id: IngestID,
+  events: Seq[IngestEvent],
+  `type`: String = "IngestEventUpdate"
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/models/IngestUpdate.scala:14]: I’m 99% sure you don’t need to add this; Circe should do it for you automatically.

(See also: how we've been reading them off SQS queues all this time!)

File: ingests_common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/tracker/IngestError.scala [hunk 6]
```diff
@@ -8,34 +8,36 @@ import uk.ac.wellcome.storage.{
...
   stored: Callback.CallbackStatus,
   update: Callback.CallbackStatus
-) extends IngestStoreError
+) extends StateConflictError
 
 case class MismatchedVersionUpdateError(
   stored: BagVersion,
   update: BagVersion
-) extends IngestStoreError
+) extends StateConflictError
```
Comment (alexwlchan) [inline ingests_common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/tracker/IngestError.scala:35]: 👍 nice

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala [hunk 7]
```diff
@@ -91,6 +111,148 @@ class IngestsTrackerApiFeatureTest
     }
   }
 
+  describe("PATCH /ingest/:id") {
+    val ingest = createIngestWith(
+      createdDate = Instant.now(),
+      events = Seq(createIngestEvent, createIngestEvent).sortBy {
+        _.createdDate
+      },
```
Comment (alexwlchan) [inline ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:119]: Any reason to sort these?

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala [hunk 8]
```diff
@@ -91,6 +111,148 @@ class IngestsTrackerApiFeatureTest
...
+        )
+
+        whenPatchRequestReady(path, updateSucceededEntity) { response =>
+          response.status shouldBe StatusCodes.OK
+          whenPatchRequestReady(path, updateFailedEntity) { response =>
+            response.status shouldBe StatusCodes.Conflict
+          }
+        }
+      }
+    }
```
Comment (alexwlchan) [inline ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:208]: Suggestion: put a succeeded ingest straight into the app at test start, i.e.

```scala
val succeededIngest = ingest.copy(status = Succeeded)

withConfiguredApp(Seq(succeededIngest)) { ingestTracker =>
  ...
}
```

File: ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala (hunk 7 above)
Comment (kenoir) [reply ingests_tracker/src/test/scala/uk/ac/wellcome/platform/storage/ingests_tracker/IngestsTrackerApiFeatureTest.scala:119]: This is copy/pasted from the ingests tests - I presume this is intended to create realistically date ordered events. I don't think we rely on the order in these tests.

File: common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/models/IngestUpdate.scala (hunk 5 above)
Comment (kenoir) [reply common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/models/IngestUpdate.scala:14]: It's puzzling - the sns/sqs mechanism uses `IndividualMessageSender` which uses the `toJson` method from the JSON lib, that however does not seem to work if I use `toJson` to serialise in the tests.

Comment (kenoir) [reply common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/models/IngestUpdate.scala:14]: You are right though - this must be possible, i'd much rather not add the extra field to IngestUpdate.

I'll try again.

Comment (alexwlchan) [reply common/src/main/scala/uk/ac/wellcome/platform/archive/common/ingests/models/IngestUpdate.scala:14]: Try importing `JsonUtil._`, I think that brings in the special code.

**Comment by kenoir** [comment]
There is some issue in CI with the bag_versioner tests - I cannot reproduce it locally. I'm guessing it may be some kind of caching issue.

**Pull request merged**